stm32/ltdc/dsi: correct properly use the pixel-format properties from ltdc node and mipi-dsi panel

avolmat-st opened the pull request:
So far the pixel-format properties from within the ltdc node is being set (mandatory in the yaml) but never used within the driver.
This is described in the yaml as the format used by the panel of the display-controller.
As a matter of fact, currently the LTDC controller only support RGB888 as output format of the controller (not to be confused with the format of data within the framebuffers, aka formats of the input layers of the LTDC).

The two first commits of this PR correct this, by checking that the pixel-format is well RGB888 and correct all formats written in dts.

Last commit correct the way MIPI-DSI is configured. Currently, the DSI format was set by looking at the LTDC format (which is in fact LTDC layer format) in order to set the ColorCoding of the MIPI-DSI. This is not correct since the ColorCoding has to be based on the format supported (and set) to the DSI panel.

Commit:
display: stm32-ltdc: clarify pixel-format property of lcd_controller

So far the LTDC driver has not be using the pixel-format property
defined in the lcd_controller bindings. This property describes the
format of the panel connected to the LTDC output. It is not to be
confused with the format of the LTDC layers (aka format of data stored
into the framebuffer). The LTDC accepts various input formats while in
most cases only RGB888 is supported for output. RGB666 is also supported
by enabling Dithering however it is not supported currently by the
driver.
Enforce verification of the pixel-format given to the LTDC device-tree
by checking that it matches with the platform supported formats.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: stm32: correct LTDC panel pixel-format to RGB888

LTDC pixel-format property has been wrongly set to various values
so far without being actually used by the LTDC driver itself.
This property describes the format of the output of the LTDC
(aka format of data between the LTDC and a panel or between
the LTDC and the MIPI-DSI block). Currently only RGB888 is
supported.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
drivers: mipi_dsi: configure DSI host using pixelformat from panel

Use the pixelformat information provided by the DSI panel in order
to control the color coding of the DSI host.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

danieldegrasse reviewed (commented):
I guess the panel internally converts the input data (regardless of format) to RGB888? Then the MIPI-DSI block can convert it to RGB-565 or similar?

Seems fine provided that the `display_get_capabilities` API is reporting the expected input buffer format

etienne-lms reviewed (commented):
Commit "display: stm32-ltdc: clarify pixel-format property of lcd_controller"
should come after commit "boards: stm32: correct LTDC panel pixel-format to RGB888"
in your series unless what build would fail on that commit.

avolmat-st commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -561,6 +561,11 @@ static DEVICE_API(display, stm32_ltdc_display_api) = {
+/* LTDC supports RGB888 and RGB666 for output however only RGB_888 is supported for now */
+#if DT_INST_PROP(0, pixel_format) != PANEL_PIXEL_FORMAT_RGB_888
+#error "Only RGB_888 is supported as a LTDC output (aka panel or mipi-dsi input format)"
+#endif
```
PANEL_PIXEL_FORMAT_RGB888 is related to the DSI format, while CONFIG_STM32_LTDC_* refer to the input format of the LTDC hence they can be different.

JarmouniA commented:
> As a matter of fact, currently the LTDC controller only support RGB888 as output format of the controller (not to be confused with the format of data within the framebuffers, aka formats of the input layers of the LTDC).

This should be clarified in the description of the `pixel-format` property in the LTDC binding, because the base description is not that.
https://github.com/zephyrproject-rtos/zephyr/blob/9463d9a51d9cb1094bf98ef437a39850a7b5705d/dts/bindings/display/lcd-controller.yaml#L13

JarmouniA commented:
> to me this is related to the pixel sent to the panel and not the data given to the controller itself.

I agree

> In the modification I am doing, I am using this pixel-format property for the format of data sent to the panel, as written in the YAML.

Okey, but in the case where LTDC is sending data to DSI block instead of the panel, the format set for LTDC can differ from the one set for the DSI device (panel controller), right? So the default binding description is not quite accurate in that case.

Commit:
dts: bindings: display: clarify lcd-controller pixel-format

Clarify the pixel-format property of a lcd-controller in order to
indicate that this is the pixel-format of data generated by the
lcd-controller, not only for a panel attached to this controller
but also for another device such as interface converter which
would be receiving the data from the lcd-controller.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

JarmouniA commented on dts/bindings/display/lcd-controller.yaml:
```diff
@@ -10,5 +10,6 @@ properties:
-      Initial Pixel format for panel attached to this controller.
-      See dt-bindings/display/panel.h for a list
+      Initial Pixel format for panel or adapter attached to this controller.
+      This can be directly a panel but can also be an interface converter.
```
```suggestion
```

akscram commented:
@avolmat-st, I ran into the assert implemented in this PR. This assert checks that `pixel-format` has to be set to `PANEL_PIXEL_FORMAT_RGB_888`.

I have a display module with the ST7789VW display controller. This controller is connected to SPI and LTDC through DPI parallel RGB interface. It's configured through SPI during its initialization phase to work through parallel RGB interface. The RGB interface is driven by LTDC. The display controller and LTDC both have to be configured in the RGB 565 pixel format mode. This means that LTDC actually outputs RGB 565 through the DPI RGB interface.

The initial configured which I tried is the following (non-related options are omitted):
```ini
# Display
CONFIG_DISPLAY=y

#MIPI
CONFIG_MIPI_DBI_SPI_3WIRE=n
CONFIG_ST7789V_RGB565=y
```
And the following in DTS (non-related parts are omitted):
```c
/ {
    mipi_dbi {
        compatible = "zephyr,mipi-dbi-spi";
        spi-dev = <&spi1>;
        status = "okay";
        spi-dev = <&spi1>;
        ...
        st7789vw: st7789vw@0 {
            compatible = "sitronix,st7789v";
            status = "okay";
            mipi-mode = "MIPI_DBI_MODE_SPI_4WIRE";
            ...
        };
};
&ltdc {
    ...
    status = "okay";
    display-controller = <&st7789vw>;
    width = <240>;
    height = <320>;
    pixel-format = <PANEL_PIXEL_FORMAT_RGB_565>;
    ...
 };
```
However, a build failed with the following assert:
```
zephyr/drivers/display/display_stm32_ltdc.c:585:2: error: #error "Only RGB_888 is supported as a LTDC output (aka panel or mipi-dsi input format)"
```
This assert is confusing because the output format in my case is actually RGB 565 and LTDC supports this format through the RGB interface.

I had to either remove the assert, which turned to be working well, or I had to add the following configuration option:
```
CONFIG_STM32_LTDC_RGB565=y
```
And set `pixel-format` in my board's DTS to `<PANEL_PIXEL_FORMAT_RGB_888>` which is confusing because it contradicts to the actual output format. This assert seems to be not-needed at least when LTDC outputs through DPI RGB interface. It could be needed when LTDC outputs to MIPI-DSI host - this I don't know. What do you think?
